### PR TITLE
[XLA:GPU][AutoSharding] Enable autosharding in GPU compiler compilation

### LINKED ISSUE #13612: How to enable AutoSharding on GPU?
## Background
I am currently exploring the PyTorch/XLA autosharding functionality on GPU platforms. It appears that GPU autosharding is not yet supported in the open-source version as per the code snippet linked below:
https://github.com/openxla/xla/blob/a921eacedc9796c26e3a88858662d05cd14a20a5/xla/service/gpu/gpu_compiler.cc#L553-L557

Further examination reveals that the **AutoSharding pass** is not enabled for GPUs, likely due to the use of #ifdef PLATFORM_GOOGLE preprocessor conditionals. I attempted to enable it by removing these conditional preprocessor directives:
https://github.com/openxla/xla/blob/a921eacedc9796c26e3a88858662d05cd14a20a5/xla/service/gpu/gpu_compiler.cc#L252-L254

https://github.com/openxla/xla/blob/a921eacedc9796c26e3a88858662d05cd14a20a5/xla/service/gpu/gpu_compiler.cc#L600-L627

## Attempted Solution and Compilation Issues
After modifying the code, I proceeded to recompile, aiming to generate the `_XLAC.so` library. Unfortunately, I encountered a compilation error indicating a missing file:
```shell
In file included from external/xla/xla/hlo/experimental/auto_sharding/auto_sharding.h:35,
                 from external/xla/xla/service/gpu/gpu_compiler.cc:272:
external/xla/xla/hlo/experimental/auto_sharding/auto_sharding_solver.h:26:10: fatal error: xla/hlo/experimental/auto_sharding/auto_sharding.pb.h: No such file or directory
   26 | #include "xla/hlo/experimental/auto_sharding/auto_sharding.pb.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
Target //:_XLAC.so failed to build
Use --verbose_failures to see the command lines of failed build steps.
```
It seems that `auto_sharding.pb.h` might be generated from `xla/hlo/experimental/auto_sharding/auto_sharding.proto`, but it's not present in the repository. To address this, I add one line `//xla/hlo/experimental/auto_sharding:auto_sharding` between **line 3469 and line 3470** in the bazel BUILD file:

https://github.com/openxla/xla/blob/a921eacedc9796c26e3a88858662d05cd14a20a5/xla/service/gpu/BUILD#L3469-L3470

Despite this modification, I encountered a link error when attempting to build `_XLAC.so`:

```shell
ERROR: /root/stage-1/pytorch/xla/BUILD:6:10: Linking _XLAC.so failed: (Exit 1): crosstool_wrapper_driver_is_not_gcc failed: error executing command (from target //:_XLAC.so) external/local_config_cuda/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc @bazel-out/k8-dbg/bin/_XLAC.so-2.params
/usr/bin/ld: bazel-out/k8-dbg/bin/external/com_google_ortools/ortools/linear_solver/libscip_with_glop.a(lpi_glop.pic.o): in function `std::_Vector_base<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::_M_deallocate(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, unsigned long)':
/proc/self/cwd/bazel-out/k8-dbg/bin/external/com_google_ortools/ortools/linear_solver/lpi_glop.cpp:148: multiple definition of `initGlopName()'; bazel-out/k8-dbg/bin/external/com_google_ortools/ortools/linear_solver/liblinear_solver.a(lpi_glop.pic.o):/proc/self/cwd/bazel-out/k8-dbg/bin/external/com_google_ortools/ortools/linear_solver/lpi_glop.cpp:148: first defined here
......
collect2: error: ld returned 1 exit status
Target //:_XLAC.so failed to build
Use --verbose_failures to see the command lines of failed build steps.
```
This error suggests a possible issue with multiple definitions.

## Questions

1. Are there any additional BUILD flags required to successfully compile `_XLAC.so` with the AutoSharding feature enabled on GPU?
2. How can I resolve the error detailed above?

## Environments

- PyTorch/XLA verision: 2.3.0

I would appreciate **any insights or suggestions on how to enable AutoSharding for GPUs** in the PyTorch/XLA project. Thank you for your assistance.

### PR DESCRIPTION
**Pull request by lausannel**
## Objective:
This pull request aims to resolve compilation issues related to enabling GPU Auto Sharding functionality in the `auto_sharding_gpu_compiler_test.cc` file. fix #13612

## Background
When compiling the `auto_sharding_gpu_compiler_test.cc` using the command:

```shell
TF_CUDA_COMPUTE_CAPABILITIES="8.0" XLA_CUDA=1 bazel build xla/service/gpu:auto_sharding_gpu_compiler_test --config=cuda
```
While the compilation is successful, the unit test subsequently fails. Upon investigation, the failure seems to be linked to the following warning, which suggests that auto sharding is not enabled.

```shell
2024-06-19 17:44:19.985341: E xla/service/gpu/gpu_compiler.cc:552] GPU autosharding is not yet available in open source.
```
## Changes Made

1. To enable GPU Auto Sharding, I removed several conditional macros in `gpu_compiler.cc`. However, this modification led to a compilation error:
    ```shell
    xla/xla/service/gpu/BUILD:3402:11: Compiling xla/service/gpu/gpu_compiler.cc failed: (Exit 1): crosstool_wrapper_driver_is_not_gcc failed: error executing command (from target //xla/service/gpu:gpu_compiler) external/local_config_cuda/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc -MD -MF bazel-out/k8-opt/bin/xla/service/gpu/_objs/gpu_compiler/gpu_compiler.pic.d ... (remaining 519 arguments skipped)
    In file included from ./xla/hlo/experimental/auto_sharding/auto_sharding.h:37,
                     from xla/service/gpu/gpu_compiler.cc:253:
    ./xla/hlo/experimental/auto_sharding/auto_sharding_solver.h:25:10: fatal error: xla/hlo/experimental/auto_sharding/auto_sharding.pb.h: No such file or directory
       25 | #include "xla/hlo/experimental/auto_sharding/auto_sharding.pb.h"
          |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    compilation terminated.
    ```

2. Upon reviewing the dependencies listed in the `BUILD` file for the `gpu_compiler` module, I discovered that the auto sharding module was not included. So I added the missing auto sharding module to the dependencies of the GPU compiler in the `BUILD` file.

## Results

Post-modification, the unit test now passes without errors, confirming that the changes fix the related issues.

### CODE REVIEW

**Review by pratikfegade — APPROVED** [review]
LGTM

**Review by ddunl — APPROVED** [review]

**Comment by ddunl** [comment]
Ok, trying to submit my patch reminded me of why this is difficult: XLA builds fine, but TensorFlow doesn't like including `auto_sharding` for similar reasons - we get a linker issue. 

```
ERROR: /tmpfs/src/copybara/tensorflow/BUILD:1318:21: Linking tensorflow/libtensorflow_cc.so.2.18.0 failed: (Exit 1): clang failed: error executing command (from target //tensorflow:libtensorflow_cc.so.2.18.0) 
  (cd /root/.cache/bazel/_bazel_root/f3e77e4f4fa3130aa77359852003683c/execroot/org_tensorflow && \
  exec env - \
    CLANG_CUDA_COMPILER_PATH=/usr/lib/llvm-17/bin/clang \
    CUDA_TOOLKIT_PATH=/usr/local/cuda-12.3 \
    DOCKER_CACHEBUSTER=1707170219221307012 \
    GCC_HOST_COMPILER_PATH=/dt9/usr/bin/gcc \
    LD_LIBRARY_PATH=/usr/local/cuda/lib64:/usr/local/cuda/extras/CUPTI/lib64 \
    PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/go/bin \
    PWD=/proc/self/cwd \
    PYTHON_BIN_PATH=/usr/bin/python3 \
    PYTHON_LIB_PATH=/usr/lib/tf_python \
    TF2_BEHAVIOR=1 \
    TF_CUDA_CLANG=1 \
    TF_CUDA_VERSION=12 \
    TF_CUDNN_VERSION=8 \
  /usr/lib/llvm-17/bin/clang @bazel-out/k8-opt/bin/tensorflow/libtensorflow_cc.so.2.18.0-2.params)
# Configuration: 232561fbd43bb8825b77a11beb284b11b92305c74aea43e1997d48184e91a3ca
# Execution platform: @sigbuild-r2.17-clang_config_platform//:platform
ld.lld: error: duplicate symbol: initGlopName()
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(initGlopName()) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/liblinear_solver.pic.a
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(.text._Z12initGlopNamev+0x0) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/libscip_with_glop.pic.a

ld.lld: error: duplicate symbol: SCIPlpiGetSolverName
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(SCIPlpiGetSolverName) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/liblinear_solver.pic.a
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(.text.SCIPlpiGetSolverName+0x0) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/libscip_with_glop.pic.a

ld.lld: error: duplicate symbol: SCIPlpiGetSolverDesc
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(SCIPlpiGetSolverDesc) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/liblinear_solver.pic.a
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(.text.SCIPlpiGetSolverDesc+0x0) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/libscip_with_glop.pic.a

ld.lld: error: duplicate symbol: SCIPlpiGetSolverPointer
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(SCIPlpiGetSolverPointer) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/liblinear_solver.pic.a
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(.text.SCIPlpiGetSolverPointer+0x0) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/libscip_with_glop.pic.a

ld.lld: error: duplicate symbol: SCIPlpiSetIntegralityInformation
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(SCIPlpiSetIntegralityInformation) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/liblinear_solver.pic.a
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(.text.SCIPlpiSetIntegralityInformation+0x0) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/libscip_with_glop.pic.a

ld.lld: error: duplicate symbol: SCIPlpiHasPrimalSolve
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(SCIPlpiHasPrimalSolve) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/liblinear_solver.pic.a
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(.text.SCIPlpiHasPrimalSolve+0x0) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/libscip_with_glop.pic.a

ld.lld: error: duplicate symbol: SCIPlpiHasDualSolve
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(SCIPlpiHasDualSolve) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/liblinear_solver.pic.a
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(.text.SCIPlpiHasDualSolve+0x0) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/libscip_with_glop.pic.a

ld.lld: error: duplicate symbol: SCIPlpiHasBarrierSolve
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(SCIPlpiHasBarrierSolve) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/liblinear_solver.pic.a
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(.text.SCIPlpiHasBarrierSolve+0x0) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/libscip_with_glop.pic.a

ld.lld: error: duplicate symbol: SCIPlpiCreate
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(SCIPlpiCreate) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/liblinear_solver.pic.a
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(.text.SCIPlpiCreate+0x0) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/libscip_with_glop.pic.a

ld.lld: error: duplicate symbol: SCIPlpiChgObjsen
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(SCIPlpiChgObjsen) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/liblinear_solver.pic.a
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(.text.SCIPlpiChgObjsen+0x0) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/libscip_with_glop.pic.a

ld.lld: error: duplicate symbol: SCIPlpiFree
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(SCIPlpiFree) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/liblinear_solver.pic.a
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(.text.SCIPlpiFree+0x0) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/libscip_with_glop.pic.a

ld.lld: error: duplicate symbol: SCIPlpiLoadColLP
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(SCIPlpiLoadColLP) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/liblinear_solver.pic.a
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(.text.SCIPlpiLoadColLP+0x0) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/libscip_with_glop.pic.a

ld.lld: error: duplicate symbol: SCIPlpiAddCols
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(SCIPlpiAddCols) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/liblinear_solver.pic.a
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(.text.SCIPlpiAddCols+0x0) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/libscip_with_glop.pic.a

ld.lld: error: duplicate symbol: SCIPlpiAddRows
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(SCIPlpiAddRows) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/liblinear_solver.pic.a
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(.text.SCIPlpiAddRows+0x0) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/libscip_with_glop.pic.a

ld.lld: error: duplicate symbol: SCIPlpiDelCols
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(SCIPlpiDelCols) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/liblinear_solver.pic.a
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(.text.SCIPlpiDelCols+0x0) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/libscip_with_glop.pic.a

ld.lld: error: duplicate symbol: SCIPlpiDelColset
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(SCIPlpiDelColset) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/liblinear_solver.pic.a
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(.text.SCIPlpiDelColset+0x0) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/libscip_with_glop.pic.a

ld.lld: error: duplicate symbol: SCIPlpiDelRows
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(SCIPlpiDelRows) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/liblinear_solver.pic.a
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(.text.SCIPlpiDelRows+0x0) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/libscip_with_glop.pic.a

ld.lld: error: duplicate symbol: SCIPlpiDelRowset
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(SCIPlpiDelRowset) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/liblinear_solver.pic.a
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(.text.SCIPlpiDelRowset+0x0) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/libscip_with_glop.pic.a

ld.lld: error: duplicate symbol: SCIPlpiClear
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(SCIPlpiClear) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/liblinear_solver.pic.a
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(.text.SCIPlpiClear+0x0) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/libscip_with_glop.pic.a

ld.lld: error: duplicate symbol: SCIPlpiChgBounds
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(SCIPlpiChgBounds) in archive bazel-out/k8-opt/bin/external/com_google_ortools/ortools/linear_solver/liblinear_solver.pic.a
>>> defined at lpi_glop.cpp
>>>            lpi_glop.pic.o:(.text.SCIPlpiChgBounds
```

**Comment by lausannel** [comment]
While compiling PyTorch/XLA, I encountered a similar issue likely related to ortools and SCIP.

Both `linear_sovler` and `scip_with_glop` module try to compile`lpi_glop.cpp` file, as the following code snippets indicate.

https://github.com/google/or-tools/blob/5425dedcfbb22cb74c636c1374a9b5ad684b1eb5/ortools/linear_solver/BUILD.bazel#L71-L82

https://github.com/google/or-tools/blob/5425dedcfbb22cb74c636c1374a9b5ad684b1eb5/ortools/linear_solver/BUILD.bazel#L192-L199

When compiling `_XLAC.so` in PyTorch/XLA, the dynamic lib will try to link both modules and casue a multiple definition error similar to `libtensorflow_cc.so`.

One workaround I found was to delete the dependency of `lpi_glop.cpp` in the `linear_sovler` module.

https://github.com/google/or-tools/blob/5425dedcfbb22cb74c636c1374a9b5ad684b1eb5/ortools/linear_solver/BUILD.bazel#L82

I recommend this fix as a potential solution but am open to any alternative suggestions. Thank you.

**Comment by ddunl** [comment]
I think this is a good fix, unfortunately the wrinkle we are going to run into is that we need to update XLA/TF's version of protobuf in order to upgrade our version of ortools, and this is difficult due to the same issues that causes the linking issues in we see in TF with this patch. Probably sooner than fixing TF's problems, we will be able to have a separate pin in XLA for protobuf, but I'm not sure I can give a fine grained timeline for that at the moment. I'll raise the issues with TF again internally, I'd really like to get this fixed

**Comment by lausannel** [comment]
Hi @ddunl I have updated the ortools patch file of ortools in the latest commit as well. Maybe we can use this solution for now and revisit the topic once we're ready to upgrade to a newer version of ortools.

**Review by ddunl — APPROVED** [review]

File: xla/service/gpu/BUILD
Comment (penpornk) [inline xla/service/gpu/BUILD:3619]: Just a heads up that we made some additional changes internally and moved this dependency to a normal list (outside of `if_gpu_is_configured`).

**Review by penpornk — COMMENTED** [review]
@lausannel 33 TF GPU unit tests failed with this PR. Could you please help take a look?
https://btx.cloud.google.com/invocations/a5e39c47-8d51-4e52-9773-60f6d8164a8c/log

(This is from a corresponding TF PR https://github.com/tensorflow/tensorflow/pull/70360)

**Comment by lausannel** [comment]
@penpornk Hi, we have invested some time in investigating this issue and, unfortunately, have not yet identified an ideal solution. Could you kindly assist us in evaluating this matter? Below are some findings that we hope you will find useful for reference.

**Reproduction Steps**:
1. Clone the TensorFlow repository using `git clone`.
2. Start the docker with the following command (root privileges may be required):
   ```shell
   TFCI=py311,linux_x86_cuda,public_cache,disk_cache ci/official/wheel.sh
   ```
3. Enter the container using `docker exec -it tf bash` and execute the following command to build the tests:
   ```
   bazel build //tensorflow/compiler/tests:conv2d_test_gpu  --repo_env=TF_PYTHON_VERSION=3.11 --config release_gpu_linux --config tf_public_cache --disk_cache=build_output/cache
   ```
4. Run the conv2d_test_gpu test with `gdb`:
    ```
    gdb --args python ./bazel-out/k8-opt/bin/tensorflow/compiler/tests/conv2d_test_gpu.runfiles/org_tensorflow/tensorflow/compiler/tests/conv2d_test_gpu --test_device=XLA_GPU --types=DT_HALF
    ```
This triggers a core dump, and here is the call stack:

image

As observed, the final core dump occurs within a Protobuf-related function, and we have conducted further investigations accordingly.

The direct cause for this issue can be traced to the file `hlo_algorithm_denylist.cc`, where protobuf attempts to parse a constant string.

https://github.com/tensorflow/tensorflow/blob/e17d963ea1f64ca49f9574fb90dae12c46a1494b/third_party/xla/xla/service/gpu/hlo_algorithm_denylist.cc#L176-L177

In `message.cc` within protobuf, the type pointer being passed is an invalid pointer, leading to a segmentation fault occurring during the subsequent search process.

https://github.com/protocolbuffers/protobuf/blob/90b73ac3f0b10320315c2ca0d03a5a9b095d2f66/src/google/protobuf/message.cc#L278-L283

The emergence of the invalid pointer is attributed to protobuf using an illegal pointer for `message_type()` during the `ConsumeField` operation, below is the related code snippet.

https://github.com/protocolbuffers/protobuf/blob/90b73ac3f0b10320315c2ca0d03a5a9b095d2f66/src/google/protobuf/text_format.cc#L631-L632

https://github.com/protocolbuffers/protobuf/blob/90b73ac3f0b10320315c2ca0d03a5a9b095d2f66/src/google/protobuf/text_format.cc#L717-L719


https://github.com/protocolbuffers/protobuf/blob/90b73ac3f0b10320315c2ca0d03a5a9b095d2f66/src/google/protobuf/generated_message_reflection.cc#L1949-L1961

Further investigation revealed that this issue might stem from the type of xla.gpu.GpuBackendConfig type being null during the building of gpu_autotuning.proto. This problem is specifically manifested in the following code:

https://github.com/protocolbuffers/protobuf/blob/90b73ac3f0b10320315c2ca0d03a5a9b095d2f66/src/google/protobuf/descriptor.cc#L6437-L6441

The size of the `symbols_by_name_` hash set unexpectedly decreases while symbols are progressively being added. As a result, the corresponding symbol for xla.gpu.GpuBackendConfig cannot be located during the build process, thereby leading to the eventual illegal pointer.

https://github.com/protocolbuffers/protobuf/blob/90b73ac3f0b10320315c2ca0d03a5a9b095d2f66/src/google/protobuf/descriptor.cc#L1506-L1511

We appreciate any insights or suggestions you might have.

**Comment by ddunl** [comment]
My suspicion is that is triggering what we try to avoid in XLA with `proto_cc_impl` targets (see impl of xla_cc_test, tf_proto_library). We do this to work around how TensorFlow builds, and I expect we'll have to do something similar for ortools. The bad news is that this is really quite difficult to do I think for a project that I'm sure doesn't want to coordinate with TensorFlow in this way....

**Comment by dimitar-asenov** [comment]
@penpornk @ddunl Hi, I'm not clear on what the state of this PR is at the moment. Do you have a suggestion for how this can be unblocked?

**Comment by penpornk** [comment]
@dimitar-asenov I was simply trying to merge this PR. I'll defer to @ddunl for his expertise. :)

**Comment by ddunl** [comment]
I think for now it's still pending passing TF tests. If TF didn't exist I would be happy to merge as is

**Comment by sgerrard** [comment]
@lausannel and @ddunl Checking in again on the status of this PR. Is there a path forward to getting TF tests to pass?

**Comment by ddunl** [comment]
I think the path forward is dependent on work that is ongoing to clean up TF's build. I believe the same problems in upgrading TF's (and therefore XLA's) protobuf version are being actively worked on, so we should reevaluate this PR after that happens.

**Comment by ddunl** [comment]
I'm going to close this for now. I know there is some ongoing work that will unblock our proto upgrade which may make the path forward on this easier, at which point we should reopen. Sorry for the difficulty here.

**Pull request closed without merging**
